### ISSUE TITLE
[css-position-3] Absolute positioning - Is the new inset & auto-size behaviour web-compatible?

### ISSUE THREAD

**Issue by dshin-moz** [w3c/csswg-drafts#11195]
Implementing `justify-self` & `align-self` for abs-positioned elements, we ran into:

* https://bugzilla.mozilla.org/show_bug.cgi?id=1930427
* https://bugzilla.mozilla.org/show_bug.cgi?id=1930850

Which has something to the effect of:

```
<!DOCTYPE html>
<style>
.abs {
  position: absolute;
  align-self: stretch;
  height: auto;
  top: auto;
  bottom: auto;
  background: lime;
}
</style>
<div class="abs">Test</div>
```

`align-self: stretch` should be a no-op (At least for absolute elements not within flex/grid context) in the old behaviour, but **seems to be used by websites in the wild anyway** in the linked bugs:

* [Virustotal](https://www.virustotal.com/gui/home/upload): `<uno-navbar>`'s shadow-root child
* [SAP](https://api.sap.com/api/OP_WAREHOUSEORDER_0001/resource/Warehouse_Order): `opblock-summary-method`

As per the new [absolute-positioning steps](https://drafts.csswg.org/css-position-3/#abspos-layout):

1. Calculate inset-modified containing block: 
    * [Inset resolution](https://drafts.csswg.org/css-position-3/#resolving-insets): `stretch`'s [fallback alignment](https://drafts.csswg.org/css-align-3/#valdef-align-content-stretch) is `start`, so take the first case - static position is at `top: 0`, `bottom` is just set to 0.
1. [Resolving sizes](https://drafts.csswg.org/css-position-3/#abspos-auto-size): Definite available size is viewport, no min/max height is given, and since self-alignment is `stretch`, **auto size is stretch size**.
1. Auto margins: N/A
1. Alignment: Fits exactly in the available space

As per the [old steps](https://drafts.csswg.org/css-position-3/#abs-non-replaced-height)

1. Everything is `auto`, margins zeroed out, `top` takes its static position value, which is zero
2. Height is [Auto heights for block formatting context roots](https://www.w3.org/TR/CSS2/visudet.html#root-height) - Distance between top of the topmost line box and bottom of the bottommost line box, so **auto size is basically a shrink-fit size**.
3. Solve for `bottom`, taking up whatever empty space

**Comment by emilio** [comment]
@dshin-moz you're probably missing a `position: absolute` in your test-case as well, right?

cc @fantasai @tabatkins

**Comment by dshin-moz** [comment]
Fixed. Thanks!

**Comment by Loirooriol** [comment]
Better add some background to the testcase to be able to see the size.
2 broken sites found in 6 days is not looking very good... Why are they using `stretch` if they don't want to stretch :facepalm:

**Comment by Loirooriol** [comment]
Actually, the default is `align-self: auto`, which behaves as `normal`, which behaves as `stretch`.

So this means that you should get this behavior even without `align-self: stretch`, which is obviously wrong.

**Comment by dshin-moz** [comment]
You're right that `align-self: auto` -> `normal`, [but then](https://drafts.csswg.org/css-position-3/#abspos-auto-size):

> Or if it is [normal](https://drafts.csswg.org/css-align-3/#valdef-align-self-normal) [...] has no [auto](https://drafts.csswg.org/css-position-3/#valdef-top-auto) [inset](https://drafts.csswg.org/css-position-3/#inset-properties) in the relevant axis
> Its [automatic size](https://drafts.csswg.org/css-sizing-3/#automatic-size) is its [stretch-fit size](https://drafts.csswg.org/css-sizing-3/#stretch-fit-size).

So given that the testcase has auto insets, it ends up being content-fit size, I think

**Comment by bfgeek** [comment]
So FWIW chrome didn't change static-pos'd abspos for this reason. We had a look at it, and determined that it wasn't web-compatible (sorry we should have filed an issue at that time).

**Comment by Loirooriol** [comment]
Yeah, in Servo we tried a few testcases on major browsers, and came to the conclusion that the automatic size of an abspos needs to be `fit-content` when one or both insets are `auto`, even if the self-alignment property is `stretch`.
https://github.com/servo/servo/blob/da4ea0f0965414733f0352e76ff416b1ded75a08/components/layout_2020/positioned.rs#L821,L829

Authors can still use `height: stretch` if they really want to stretch.

**Comment by Loirooriol** [comment]
Actually, it has already been fixed in the spec (#11902): https://drafts.csswg.org/css-position-3/#abspos-auto-size

<blockquote>

The [automatic size](https://drafts.csswg.org/css-sizing-3/#automatic-size) of an [absolutely positioned box](https://drafts.csswg.org/css-position-3/#absolute-position) is resolved against its [inset-modified containing block](https://drafts.csswg.org/css-position-3/#inset-modified-containing-block) as follows:

<dl>
<dt>

If its [self-alignment property](https://drafts.csswg.org/css-align-3/#self-alignment-properties) in the relevant axis is [stretch](https://drafts.csswg.org/css-align-3/#valdef-align-self-stretch) and neither of its [inset properties](https://drafts.csswg.org/css-position-3/#inset-properties) nor [margins](https://drafts.csswg.org/css-box-4/#margin) in that axis are auto

</dt>
<dt>

Or if it is [normal](https://drafts.csswg.org/css-align-3/#valdef-align-self-normal) and the box is [non-replaced](https://drafts.csswg.org/css-display-4/#non-replaced), not a [table wrapper box](https://drafts.csswg.org/css-tables-3/#table-wrapper-box), and has no [auto](https://drafts.csswg.org/css-position-3/#valdef-top-auto) [inset](https://drafts.csswg.org/css-position-3/#inset-properties) in the relevant axis

</dt>
<dd>

Its [automatic size](https://drafts.csswg.org/css-sizing-3/#automatic-size) is its [stretch-fit size](https://drafts.csswg.org/css-sizing-3/#stretch-fit-size).

</dd>
<dt>

Otherwise

</dt>
<dd>

Its [automatic size](https://drafts.csswg.org/css-sizing-3/#automatic-size) is its [fit-content size](https://drafts.csswg.org/css-sizing-3/#fit-content-size).

</dd>
</dl>

</blockquote>

Well, the "nor [margins](https://drafts.csswg.org/css-box-4/#margin) in that axis are auto" is wrong: browsers still allow stretching.

**Comment by Loirooriol** [comment]
I think it's completely nonsensical that `place-self: stretch` does NOT stretch when `place-self: normal` DOES stretch!

```html
<!DOCTYPE html>
<style>
.container {
  position: relative;
  width: 100px;
  height: 100px;
  border: solid;
}
.abspos {
  position: absolute;
  inset: 0;
  margin: auto;
  border: solid magenta;
}
</style>
<div class="container">
  <div class="abspos" style="place-self: normal">foo bar</div>
</div>
<div class="container">
  <div class="abspos" style="place-self: stretch">foo bar</div>
</div>
```

![](https://github.com/user-attachments/assets/6ab4b9ca-f4a7-46a9-b5e4-83768cb39f09)

Sure, auto margins win in other layouts, but they win over both `normal` and `stretch` alignments.
I prefer abspos layout to be self-consistent rather than making only a few corner cases consistent with other layouts.

**Comment by fantasai** [comment]
> I prefer abspos layout to be self-consistent rather than making only a few corner cases consistent with other layouts.

If we make `stretch` win over auto margins, then it becomes inconsistent with all of the other alignment values which do not win over auto margins.

If we make all alignment values win over auto margins, then we make alignment vs margins inconsistent between abspos and all other layout models (not just a few corner cases).

The last option is the one we have here, which is that auto margins consistently win over alignment values, except that `normal` behavior is weird.

There is no fully consistent behavior, so the question is, which is the one we want?

**Comment by Loirooriol** [comment]
> it becomes inconsistent with all of the other alignment values which do not win over auto margins.

You can see it in a different way. Ignoring float details from [#9963](https://github.com/w3c/csswg-drafts/issues/9963),
 - Self-alignment properties can have 2 different effects:
   - They can force the automatic size to be `fit-content`
   - They align the margin box
 - Auto margins can have 2 different effects:
   - They can force the automatic size to be `fit-content`
   - They align the border box by modifying the size of the margin box

So it's not really a matter of which wins. If neither `stretch` alignment nor `auto` margins force a `fit-content` size, then neither alignment effect will be observable. If other alignment values force a `fit-content` size, then auto margins can take the free space, and then aligning the margin box is not observable. Doesn't seem self-inconsistent to me.
> inconsistent between abspos and all other layout models

It would be consistent with block-level layout, where `auto` margins do not prevent `justify-self: normal/stretch` from stretching.

For other values it needs to be confirmed in #12102, but I would say that again abspos and block-level layouts should be consistent.

**Comment by fantasai** [comment]
> If neither stretch alignment nor auto margins force a fit-content size, then neither alignment effect will be observable.

This is not true. The box may have a non-stretch-fit size, in which case alignment is observable.

**Comment by bfgeek** [comment]
@fantasai @Loirooriol - Is it correct under the above resolution, tables should have the old behaviour? E.g.

https://www.software.hixie.ch/utilities/js/live-dom-viewer/?saved=13785

```
<!DOCTYPE html>
<style>
.cb {
  width: 100px;
  height: 100px;
  display: inline-block;
  position: relative;
  border: solid 3px;
}
table {
  position: absolute;
  min-width: 20px;
  min-height: 20px;
  background: lime;
  margin: auto;
  inset: 0;
}
</style>
<div class="cb">
  <table style="align-self: normal; justify-self: normal;"></table>
</div>
<div class="cb">
  <table style="align-self: stretch; justify-self: stretch;"></table>
</div>
```

**Comment by Loirooriol** [comment]
Oh, yeah, I failed to consider tables and replaced elements when summarizing the proposed resolution, but I guess they should be covered too.

So basically https://drafts.csswg.org/css-position/#abspos-auto-size should be changed as follows:

<blockquote>

The [automatic size](https://drafts.csswg.org/css-sizing-3/#automatic-size) of an [absolutely positioned box](https://drafts.csswg.org/css-position-3/#absolute-position) is resolved against its [inset-modified containing block](https://drafts.csswg.org/css-position-3/#inset-modified-containing-block) as follows:

<dl>
<dt>

If its [self-alignment property](https://drafts.csswg.org/css-align-3/#self-alignment-properties) in the relevant axis is [stretch](https://drafts.csswg.org/css-align-3/#valdef-align-self-stretch) and neither of its [inset properties](https://drafts.csswg.org/css-position-3/#inset-properties) <del>nor [margins](https://drafts.csswg.org/css-box-4/#margin)</del> in that axis are auto

</dt>
<dt>

Or if it is [normal](https://drafts.csswg.org/css-align-3/#valdef-align-self-normal) and the box is [non-replaced](https://drafts.csswg.org/css-display-4/#non-replaced), not a [table wrapper box](https://drafts.csswg.org/css-tables-3/#table-wrapper-box), and has no [auto](https://drafts.csswg.org/css-position-3/#valdef-top-auto) [inset](https://drafts.csswg.org/css-position-3/#inset-properties) in the relevant axis

</dt>
<dd>

Its [automatic size](https://drafts.csswg.org/css-sizing-3/#automatic-size) is its [stretch-fit size](https://drafts.csswg.org/css-sizing-3/#stretch-fit-size).

</dd>
<dt>

Otherwise

</dt>
<dd>

Its [automatic size](https://drafts.csswg.org/css-sizing-3/#automatic-size) is its [fit-content size](https://drafts.csswg.org/css-sizing-3/#fit-content-size).

</dd>
</dl>

</blockquote>

**Comment by bfgeek** [comment]
Sounds good. Tables stretching is this case is more consistent IMO.

**Comment by tabatkins** [comment]
Thanks, updated the draft.
